src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/Compression/ZLibStreamPool.cs

```diff
@@ -0,0 +1,287 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using static System.IO.Compression.ZLibNative;
+
+namespace System.Net.WebSockets.Compression
+{
+    internal sealed class ZLibStreamPool
+    {
+        private static readonly ZLibStreamPool?[] s_pools
+            = new ZLibStreamPool[WebSocketValidate.MaxDeflateWindowBits - WebSocketValidate.MinDeflateWindowBits + 1];
+
+        /// <summary>
+        /// The default amount of time after which a cached item will be removed.
+        /// </summary>
+        private const int DefaultTimeoutMilliseconds = 60_000;
+
+        private readonly int _windowBits;
+        private readonly List<CacheItem> _inflaters = new();
+        private readonly List<CacheItem> _deflaters = new();
+        private readonly Timer _cleaningTimer;
```

Comment:
to be honest with you I am sceptical about adding another custom object pool implementation. What are the existing alternatives?

@jkotas do you have any recommendations?

Reply:
I hear you. However I failed to find any good alternatives that I can use. If you know something that I could use that would allow the same functionality, I am more than happy to use it.

Reply:
The simplest alternative is to not pool. The GC is pretty good at dealing with small objects accross the whole range of configurations.

Do you have data that show the benefits of this pool?

Reply:
@jkotas the problem ain't the GC. What we're pooling here is native memory. In the most common case (15 bits window size), this is 32KB per instance.

Reply:
Also to note, that the initial implementation didn't have this pool. It was added later after discussion with @CarnaViire and @scalablecory.

Reply:
@jkotas thanks, I didn't know that. See original discussion that lead to the addition of this pool - https://github.com/dotnet/runtime/pull/49304#discussion_r593364659.

Please talk to @scalablecory and @CarnaViire if possible, and let me know your decision.

Reply:
I think it would be reasonable to cache one instance of the buffer in the websocket instance (without the flush on timer). I would recomend simplifying this PR to do that. It will cover the most common case.

Going beyond that - caching more than one; or creating global pools - needs a lot of performance testing to prove that it works well under broad range of scenarios.

@scalablecory and @CarnaViire Thoughts?

Reply:
@jkotas My opinion is that caching an instance per socket, when context takeover is explicitly disabled, defeats the purpose of the flag.

Reply:
What are the typical realworld use cases for websocket with disabled context takeover?

I can see that it makes sense to disable context takeover for low-traffic channel where it is not worth it to keep the inflater/deflater state around between messages. Any other typical use cases?

Then I would suggest to not cache the buffer at all when the context takeover is disabled, and just depend on unmanaged memory allocator doing a decent job.

Reply:
Admittedly we're operating on an assumption that init/destroy of zlib is too expensive to happen on a per-message basis. For e.g. SignalR one message might be written hundreds of times to broadcast to multiple clients, so there can be some major burstiness to how it will be called. We could certainly benchmark if this is a strong concern @jkotas.

Another strategy we could take is to have a single object cached in a thread local. This would allow no-context-takeover case to operate efficiently in its common usage.

Otherwise, I'm comfortable keeping the current cache and adjusting later if needed. It's not clear to me that the current code will have significant contention -- I would expect most time to be spent outside of lock in compress/decompress, encryption, and I/O, and it's becoming increasingly rare to see vertical rather than horizontal scaling for web apps.

Reply:
>  It's not clear to me that the current code will have significant contention -- I would expect most time to be spent outside of lock in compress/decompress, encryption, and I/O

With the current pool implementation and large machine, the code outside the lock that operates on the pooled item will typically run many times slower since it will be contending with caches on other CPUs.

Reply:
ahh you are worried about sharing. the TLS solution would solve that.

Reply:
Right, it would solve that at the cost of potentially keeping a lot of unused memory around.

In any case, this needs performance numbers to demonstrate what it is saving.

Reply:
@jkotas I might have been optimistic about the memory and forgot that I've measured it before - see here https://github.com/dotnet/runtime/issues/31088#issuecomment-539603012. The deflate with 15 bits needs 256KB. I've modified the code to use `AppContext` switch to turn off caching and written a simple benchmark that measures the impact. I'm sending to `Stream.Null` so the measurement depicts an accurate picture of just the deflate.

``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19042
11th Gen Intel Core i7-11700 2.50GHz, 1 CPU, 16 logical and 8 physical cores
.NET Core SDK=6.0.100-preview.2.21155.3
  [Host]     : .NET Core 5.0.5 (CoreCLR 5.0.521.16609, CoreFX 5.0.521.16609), X64 RyuJIT
  Job-DAXMTF : .NET Core 5.0 (CoreCLR 42.42.42.42424, CoreFX 42.42.42.42424), X64 RyuJIT

Toolchain=CoreRun  

```
| Method | Cache |     Mean |    Error |   StdDev |   Median |  Gen 0 | Gen 1 | Gen 2 | Allocated | Allocated native memory |
|------- |------ |---------:|---------:|---------:|---------:|-------:|------:|------:|----------:|------------------------:|
|   **Send** | **False** | **47.34 μs** | **0.938 μs** | **2.155 μs** | **48.17 μs** | **0.0610** |     **-** |     **-** |     **568 B** |                **268576 B** |
|   **Send** |  **True** | **32.30 μs** | **0.058 μs** | **0.049 μs** | **32.28 μs** |      **-** |     **-** |     **-** |     **488 B** |                       **-** |

Reply:
Also I don't think the `[ThreadStatic]` path would be a good one, because it might end up wasting a lot of memory. 

I can try and improve the implementation of the current cache and use buckets per logical processor, but I don't know why I can't use `Thread.GetCurrentProcessorId` in `System.Net.WebSockets`.

Reply:
> Then I would suggest to not cache the buffer at all when the context takeover is disabled, and just depend on unmanaged memory allocator doing a decent job.

> In any case, this needs performance numbers to demonstrate what it is saving.

I totally agree with @jkotas

IMO we should:

* remove the cache from this PR and merge a simple version
* author new benchmarks and add them to https://github.com/aspnet/benchmarks, ask @sebastienros to include them in the Power BI report
* run the benchmarks for various configs ((Windows, Linux)x(12, 28, 42, 56 cores)x(x64, ARM64). This can be done only by someone who works at MS and has access to VPN. 
* based on findings, fix the perf if needed

Reply:
@adamsitnik this sounds good to me. Also I think that the most common use case will be with `context takeover` anyway. In order to avoid doing things twice, please discuss this with @scalablecory and let me know.

Reply:
One other thing that comes to mind is that I can put the cache behind am `AppContext Switch`, which can be turned on/off by default as you decide. This will allow for easier testing and if results prove that it's net negative, we can remove it without anyone noticing.

Reply:
I'm fine with @adamsitnik's recommendation. If anyone in the path of the perf regression is affected substantially, they'll presumably yell loud enough to have it looked at later. Let's get this merged!

Reply:
> The deflate with 15 bits needs 256KB.

If somebody configures buffer that is this big, they better have to write a good amount of data. I do not think that testing this with empty stream is represenative. The test should write at least as much data as the buffer size to be represenative.

To make progress, I agree that it would be best to delete the pooling from this PR.

Reply:
Done.